src/de/packet_parser.rs

```diff
@@ -113,21 +104,31 @@ impl<'a> PacketParser<'a> {
     pub fn read_properties<'b, 'me: 'b>(&'me self) -> Result<Vec<Property<'b>, 8>, Error> {
         let mut properties: Vec<Property, 8> = Vec::new();
 
-        let properties_size = self.read_variable_length_integer()?;
-        let mut property_bytes_processed = 0;
+        let properties_size = self.read_variable_length_integer()? as usize;
 
-        while properties_size - property_bytes_processed > 0 {
-            let property = Property::parse(self)?;
-            property_bytes_processed += property.size();
+        let remainder = self.len()? - properties_size;
+
+        while self.len()? > remainder {
             properties
-                .push(property)
+                .push(Property::parse(self)?)
                 .map_err(|_| Error::MalformedPacket)?;
         }
 
-        if properties_size != property_bytes_processed {
+        if self.len()? != remainder {
             return Err(Error::MalformedPacket);
         }
 
         Ok(properties)
     }
 }
+
+pub struct IntReader<'a> {
+    parser: &'a PacketParser<'a>,
+}
+
+impl<'a> varint_rs::VarintReader for IntReader<'a> {
+    type Error = Error;
+    fn read(&mut self) -> Result<u8, Error> {
```

Comment:
This is only currently needed because the `VarintReader` trait needs `&mut self`, but our `read()` method only has access to `&self`